### DISCUSSION TITLE
question: what is used to generate the layout image?

### DISCUSSION THREAD

**Discussion by mattmarcello** [urob/zmk-config#33] [Q&A] [answered]
might be cool to add to produce this as a build artifact by default?

**Comment by urob** [comment, marked as answer]
**Update**

The layout is generated by `just draw`, which utilizes `keymap-drawer` to parse and plot the keymap. For details see #76. 

--

**Old answer**

I am currently creating the layout manually. Here's the source code if you want to modify: http://www.keyboard-layout-editor.com/#/gists/06469e3337fa76f1c6e95229f5c58522

~At some point, I am going to move over to use https://github.com/caksoylar/keymap-drawer to automate things, but I haven't found time yet to set things up to work end-to-end.~ 

There's now an example workflow (#76) showing how to automate the keymap using keymap-drawer. TL;DR: `just draw` does everything if you are using my local build environment (#74).

**Reply by angularsen** [reply]
So based on https://github.com/urob/zmk-config/discussions/68#discussioncomment-10096321 I found a way to generate this manually, that probably can be automated if the physical layout can be described in the `.keymap` file somehow?

```sh
pip install pcpp
cd config # In your zmk-config repo root
pcpp --passthru-unfound-includes your-file.keymap > flat.keymap # Flatten the keymap
```

1. Go to https://keymap-drawer.streamlit.app/
1. `Parse from ZMK keymap` > Browse file and select `flat.keymap` created above > `Parse from file`
1. You will get an error saying the physical layout must be described, add `layout:` to the yaml, for example `layout: {cols_thumbs_notation: 33333+3 3+33333}` for split with 5 columns, 3 rows and 3 thumb buttons each. [More details on layout syntax here](https://github.com/caksoylar/keymap-drawer/blob/main/KEYMAP_SPEC.md#layout).


![image](https://github.com/user-attachments/assets/3bdd039e-cf80-469d-a645-63be78ecf858)

![image](https://github.com/user-attachments/assets/0bd84325-f354-4a22-853d-bd58fd17e776)

**Reply by angularsen** [reply]
Not as pretty as the one in the README, and I don't see things like long press for HOME/END etc. But it's a start at least.

**Reply by urob** [reply]
Let me flag some work-in-progress that adds support to keymap-drawer for configs that rely on Zephyr modules (such as the `v2` branch of zmk-helpers used in this repo). 

1. [#105](https://github.com/caksoylar/keymap-drawer/pull/105) adds support for zmk-helpers-v2 when running keymap-drawer locally (currently merged but not yet released)
2. [#107](https://github.com/caksoylar/keymap-drawer/pull/107) adds support for zmk-helpers-v2 when running keymap-drawer through Github Actions (currently waiting to be merged)

Once these PRs are both merged and released, using keymap-drawer with this repo should be hopefully much simpler. For an example how to already use them now, check out @stijnveenman's [zmk-config](https://github.com/stijnveenman/zmk-config). 

Based on their work, I also did [some first exploring](https://github.com/urob/zmk-config/tree/keymap-drawer) myself to get the local workflow running. This for now requires a custom build of keymap-drawer (since #105 isn't released yet) but otherwise works fine. I haven't tried tuning the resulting images at all though...

**Reply by urob** [reply]
With some delay, I finally got the keymap-drawer workflow polished and fully automated. See this discussion for details: https://github.com/urob/zmk-config/discussions/76

**Comment by autoferrit** [comment]
has there been any effort on using keymap drawer? the nodefree repo that this uses seems to be incompatible as there is no root node in the keymap file.

**Reply by angularsen** [reply]
https://github.com/urob/zmk-config/discussions/33#discussioncomment-10119385

**Comment by urob** [comment]
> has there been any effort on using keymap drawer? the nodefree repo that this uses seems to be incompatible as there is no root node in the keymap file. 

Keymap drawer should work (it does for me). It is smart enough to include all necessary files and expand all macros. 

Two potential stumbling points:
1. Make sure to run it on the root file (eg corneish_zen.keymap). Running it in base.keymap doesn't work since base is intended to be included in another keymap file. 

2. If you include the nodefree-repo as submodule, the online version of keymap drawer will currently fail (there's already a PR that fixes this). In either case, running keymap drawer locally works

**Reply by urob** [reply]
Quick update: Since version `0.18.0` keymap-drawer also supports using `zmk-helpers` both as a Zephyr module. I added some instructions here with details:
https://github.com/urob/zmk-helpers/blob/main/docs/keymap_drawer.md
